### DISCUSSION TITLE
NaN in gradient computations

### DISCUSSION THREAD

**Discussion by giovannifereoli** [casadi/casadi#3299] [General]
Can anyone help me in understanding why I this function fails?

from mpopt import mp
import numpy as np
import casadi as ca

def dynamics(X, u, t):
    # Data
    mu = 398600.0
    R = 42241.08
    n = ca.sqrt(mu / R**3)
    Isp = 310.0
    g0 = 9.81

    # Unpacking state
    x = X[0]
    y = X[1]
    z = X[2]
    xdot = X[3]
    ydot = X[4]
    zdot = X[5]
    m = X[6]
    T = u
    Tx = T[0]
    Ty = T[1]
    Tz = T[2]
    Tnorm = ca.sqrt(Tx * Tx + Ty * Ty + Tz * Tz)

    return [
        xdot,
        ydot,
        zdot,
        3 * n**2 * x + 2 * n * ydot + Tx / m,
        -2 * n * xdot + Ty / m,
        -(n**2) * z + Tz / m,
        -Tnorm / (Isp * g0),
    ]

def running_costs(x, u, t):
    return ca.sqrt(u[0] * u[0] + u[1] * u[1] + u[2] * u[2])

def terminal_constraints(x, t, x0, t0):
    return [x[0], x[1], x[2], x[3], x[4], x[5]]

ocp = mp.OCP(n_states=7, n_controls=3, n_phases=1)
ocp.dynamics[0] = dynamics
ocp.running_costs[0] = running_costs
ocp.terminal_constraints[0] = terminal_constraints

ocp.x00[0] = [0, 200, 0, 0, 0, 0, 21000]

T = 29000
ocp.lbu[0], ocp.ubu[0] = [-T, -T, -T], [T, T, T]

ocp.validate()
mpo, post = mp.solve(ocp, n_segments=1, poly_orders=25, scheme="LGR", plot=True)
mp.plt.show()

I've tried also different dynamics functions, but every time I've this type of errors " CasADi - 2023-07-25 18:11:51 WARNING("solver:nlp_grad_f failed: NaN detected for output grad_f_x, at (row 182, col 0).") [.../casadi/core/oracle_function.cpp:326]
CasADi - 2023-07-25 18:11:51 WARNING("solver:nlp_jac_g failed: NaN detected for output jac_g_x, at nonzero index 4969 (row 103, col 181).") [.../casadi/core/oracle_function.cpp:326]
CasADi - 2023-07-25 18:11:51 WARNING("solver:nlp_jac_g failed: NaN detected for output jac_g_x, at nonzero index 4969 (row 103, col 181).") [.../casadi/core/oracle_function.cpp:326]
CasADi - 2023-07-25 18:11:51 WARNING("solver:nlp_g failed: NaN detected for output g, at (row 103, col 0).") [.../casadi/core/oracle_function.cpp:326]
CasADi - 2023-07-25 18:11:51 WARNING("solver:nlp_grad failed: NaN detected for output grad_gamma_p, at (row 0, col 0).") [.../casadi/core/oracle_function.cpp:326]
CasADi - 2023-07-25 18:11:51 WARNING("Failed to calculate multipliers") [.../casadi/core/nlpsol.cpp:798]
"

**Comment by brunoml5** [comment]
Hi! There are many problems with your model.

NaN usually happens when there are invalid math operations such as division by zero or square-root of a negative number. In square-root, avoid negative number by adding a small positive number, such as:
```
Tnorm = ca.sqrt(Tx*Tx + Ty*Ty + Tz*Tz + 0.0001)
```

Also, avoid square-root in your objective function. Just put the function squared:

```
def running_costs(x, u, t):
    return u[0] * u[0] + u[1] * u[1] + u[2] * u[2]
```

For divisions, avoid zero values by adding an small number, such as:

```
return [
        xdot,
        ydot,
        zdot,
        3 * n**2 * x + 2 * n * ydot + Tx / (m+0.001),
        -2 * n * xdot + Ty / (m+0.001),
        -(n**2) * z + Tz / (m+0.001),
        -Tnorm / (Isp * g0),
    ]
```

And maybe add a lower bound for `m` with:
```
ocp.lbx[0, 6] = 0
```

Also, it is weird to use polynomial with such a high order (25) for integration. Try using a polynomial of order 5 and make more integration steps.

I hope it helps!
